mlbench_core/optim/pytorch/optim.py

```diff
@@ -499,12 +499,290 @@ def step(self, closure=None):
+        ]
+        high_rank_tensors = [
+            (tensor, out, mem)
+            for tensor, out, mem in zip(grad_in, grad_out, memory_out)
+            if tensor.ndimension() > 1
+        ]
+
+        # We are building a rank-1 approximation of every tensor
+        # that can be interpreted as a matrix. Let the approximation be
+        # M = p q^T
+        # We are allocating consequtive memory for the p's and q's
+
+        memory_is_uninitialized = self.p_memory is None
+
+        p_total_size = 0
+        q_total_size = 0
+        for tensor, _, _ in high_rank_tensors:
+            matrix = tensor.view(tensor.shape[0], -1)
+            n, m = matrix.shape
+            rank = min(n, m, self.rank)
+            p_total_size += n * rank
+            q_total_size += m * rank
+        if self.p_memory is None:
+            self.p_memory = torch.empty(p_total_size, device=self.device)
+            self.q_memory = torch.empty(q_total_size, device=self.device)
+
+        # Find them again and make lists of pointers
+        ps = []
+        qs = []
+        p_idx = 0
+        q_idx = 0
+        for tensor, _, _ in high_rank_tensors:
+            matrix = tensor.view(tensor.shape[0], -1)
+            n, m = matrix.shape
+            rank = min(n, m, self.rank)
+            ps.append(self.p_memory[p_idx: p_idx + n * rank].view(n, rank))
+            qs.append(self.q_memory[q_idx: q_idx + m * rank].view(m, rank))
+            p_idx += n * rank
+            q_idx += m * rank
+
+        for (tensor, _, _), q, p in zip(high_rank_tensors, qs, ps):
+            matrix = tensor.view(tensor.shape[0], -1)
+            n, m = matrix.shape
+
+            if self.reuse_query and not memory_is_uninitialized:
+                # orthogonalize(q)
+                pass
+            else:
+                # Sample a query vector q
+                self.set_random(q)
+
+        for (tensor, _, _), q, p in zip(high_rank_tensors, qs, ps):
+            matrix = tensor.view(tensor.shape[0], -1)
+            torch.matmul(matrix, q, out=p)
+
+        all_reduce(self.p_memory)
+
+        # Start communicating rank 1 tensors
+        rank1_tensor_list = TensorBuffer([tensor for (tensor, _, _) in rank1_tensors])
+
+        rank1_handle = rank1_tensor_list.all_reduce(async_op=True)
+
+        for p in ps:
+            orthogonalize(p)
+
+        for p, q, (tensor, _, _) in zip(ps, qs, high_rank_tensors):
+            matrix = tensor.view(tensor.shape[0], -1)
+            torch.matmul(matrix.t(), p, out=q)
+
+        all_reduce(self.q_memory)
+        self.q_memory.data[:] /= self.n_workers
+
+        for p, q, (tensor, out, mem) in zip(ps, qs, high_rank_tensors):
+            # Set the output gradient
+            torch.matmul(p, q.t(), out=out.data[:])
+            mem.data[:] = tensor - out
+
+        rank1_handle.wait()
+        rank1_tensor_list.buffer /= self.n_workers
+        rank1_tensor_list.unpack([out for (_, out, _) in rank1_tensors])
+
+
+class TensorBuffer:
```

Comment:
should this one be used by the all optimizers?

Reply:
No, only by PowerSGD. I will put a docstring to explain this.

Reply:
it's related to the per-layer or per-model aggregation also, which we are supporting for all optimizers #61 . maybe the code could be shared?

Reply:
Ideally, this class should be moved to `mlbench_core/utils/pytorch/distributed.py` since it doesn't really have to do with optimizers. Also, a function `pack_tensors` already exists, and I'm not sure we need a whole class for it. Please check how it's used.